bundles/org.openhab.binding.iotawatt/src/main/java/org/openhab/binding/iotawatt/internal/service/FetchDataService.java

```diff
@@ -0,0 +1,145 @@
+/**
+ * Copyright (c) 2010-2024 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.iotawatt.internal.service;
+
+import java.net.URISyntaxException;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.iotawatt.internal.client.IoTaWattClient;
+import org.openhab.binding.iotawatt.internal.model.IoTaWattChannelType;
+import org.openhab.binding.iotawatt.internal.model.StatusResponse;
+import org.openhab.core.library.types.QuantityType;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Fetches data from IoTaWatt and updates the channels accordingly.
+ *
+ * @author Peter Rosenberg - Initial contribution
+ */
+@NonNullByDefault
+public class FetchDataService {
+    static final String INPUT_CHANNEL_ID_PREFIX = "input_";
+    static final String OUTPUT_CHANNEL_ID_PREFIX = "output_";
+
+    private final Logger logger = LoggerFactory.getLogger(FetchDataService.class);
+
+    private final DeviceHandlerCallback thingHandler;
+    private @Nullable IoTaWattClient ioTaWattClient;
+
+    public FetchDataService(DeviceHandlerCallback thingHandler) {
+        this.thingHandler = thingHandler;
+    }
+
+    public void setIoTaWattClient(IoTaWattClient ioTaWattClient) {
+        this.ioTaWattClient = ioTaWattClient;
+    }
+
+    /**
+     * Poll the device once without retry.
+     * Handles error cases and updates the Thing accordingly.
+     */
+    public void pollDevice() {
+        final IoTaWattClient nonNullClient = ioTaWattClient;
+        if (nonNullClient == null) {
+            thingHandler.updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.HANDLER_INITIALIZING_ERROR);
```

Comment:
I see your point and that's in fact what I tried to achieve, together with having no logic at all in `ThingHandler`.
To do this, I created an interface `DeviceHandlerCallback` so that `FetchDataService` has no knowledge of `ThingHandler` but can still update the `ThingStatus`.

When I let the exception bubble up to `ThingHandler`, I need some logic there what I can prevent that way. This allows me to test all logic with JUnit and don't need to Unit-Test the `ThingHandler` as it has no logic at all.
Do you see what I mean?

Reply:
Yes, i see. When using `FetchDataService` as an intermediate / mapping layer between the `ThingHandler` and the `IoTaWattClient` this is perfectly sane.

Maybe it has to do something with the service name and/or file structure, anyway, let's leave it as is and continue.

Reply:
Nice do we agree on advantages of decoupling from `ThingHandler`. I intentionally placed `DeviceHandlerCallback` in the same package as `FetchDataService` what I feel is the right approach. The class variable name is not ideal as `thingHandler` and I just renamed it to `deviceHandlerCallback`. To make it more clear we could even think on naming the interface `FetchDataServiceDeviceHandlerCallback`? I feel it's probably a bit overkill.

Reply:
Yeah that is fine and agree the  `FetchDataServiceDeviceHandlerCallback` is a bit too much.